### PR TITLE
Resize with resampling kernels

### PR DESCRIPTION
**Pull request by mzient**
Integrate resampling kernels with Resize Op.
Remove NewResize and all of its dependencies from Resize.
Fix tests.

### CODE REVIEW

File: dali/kernels/common/copy.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <cuda_runtime.h>
+#include "dali/kernels/tensor_view.h"
+#include "dali/kernels/backend_tags.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename Storage1, typename T1, int ndim1, typename Storage2, typename T2, int ndim2>
+void copy(const TensorView<Storage1, T1, ndim1> &out,  // NOLINT
```
Comment (jantonguirao) [inline dali/kernels/common/copy.h]: how come that the output argument is marked as const?

File: dali/kernels/common/copy.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#define DALI_KERNELS_COMMON_COPY_H_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/tensor_view.h"
+#include "dali/kernels/backend_tags.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename Storage1, typename T1, int ndim1, typename Storage2, typename T2, int ndim2>
```
Comment (jantonguirao) [inline dali/kernels/common/copy.h]: I'd change the 1, 2 suffixes with Input, Output. E.g. `StorageInput`, `StorageOutput`

File: dali/kernels/imgproc/resample.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      return cur_impl->get();
+    } else {
+      auto ptr = Impl::Create(params);
+      context.kernel_data = ptr;
+      return ptr.get();
+    }
+  }
+  static Impl *GetImpl(KernelContext &context) {
+    auto cur_impl = any_cast<ImplPtr>(&context.kernel_data);
+    if (cur_impl) {
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample.h]: this if/else seems a bit redundant

File: dali/kernels/imgproc/resample.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      return cur_impl->get();
+    } else {
+      return nullptr;
+    }
+  }
+
+  static KernelRequirements
+  GetRequirements(KernelContext &context, const Input &input, const Params &params) {
+    auto *impl = SelectImpl(context, input, params);
+    return impl->Setup(context, input, params);
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample.h]: `assert(impl != nullptr);`

File: dali/kernels/imgproc/resample/bilinear_impl.cuh [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+namespace {
+
+/// @brief Implements horizontal resampling for a custom ROI
+/// @param x0 - start column, in output coordinates
+/// @param x1 - end column (exclusive), in output coordinates
+/// @param y0 - start row
+/// @param y1 - end row (exclusive)
+template <int static_channels = -1, typename Dst, typename Src>
+__device__ void LinearHorz_Channels(
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/bilinear_impl.cuh]: comment says it implements horizontal resampling but the function name doesn't really tell me what it does

File: dali/kernels/imgproc/resample/resampling_batch.cu [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  int block = sbi.block_in_sample;
+
+  // 0 = vertical axis, 1 = horizontal axis
+  int axis = (which_pass == 0) ^ (sample.order == SeparableResamplingSetup::VertHorz);
+
+  ResamplingFilterType ftype = sample.filter_type[axis];
+  ResamplingFilter filter = sample.filter[axis];
+  int support = filter.support();
+
+  float scale = in_shape[axis] * (1.0f / out_shape[axis]);
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_batch.cu]: are we sure that `out_shape[axis]` is never 0?

File: dali/kernels/imgproc/resample/resampling_filters.cu [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+__global__ void InitGaussianFilter(ResamplingFilter filter) {
+  InitFilter(filter, [&](int i) {
+    float x = 4 * (i - (filter.num_coeffs-1)*0.5f) / (filter.num_coeffs-1);
+    return expf(-x*x);
+  });
+}
+
+inline __host__ __device__ float sinc(float x) {
+  return x ? sinf(x * M_PI) / (x * M_PI) : 1;
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_filters.cu]:
```suggestion
  return x ? sinf(x * M_PI) / (x * M_PI) : 1.0f;
```

 ?

File: dali/kernels/imgproc/resample/resampling_filters_test.cu [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+__global__ void GetFilterValues(float *out, ResamplingFilter filter, int n, float step) {
+  int i = threadIdx.x + blockIdx.x*blockDim.x;
+  if (i >= n)
+    return;
+  out[i] = filter(i*step);
+}
+
+TEST(ResamplingFilters, DISABLED_PrintFilters) {
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_filters_test.cu]: probably good idea to add expectations and enable this test

File: dali/kernels/imgproc/resample/resampling_impl.cuh [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+///
+/// The function follows different code paths for static and dynamic number of channels.
+/// For the dynamic, the innermost loop goes over filter taps, which eliminates the neeed
+/// for thread-local memory to store intermediate sums. This allows processing arbitrary
+/// number of channels.
+/// For static number of channels, the run-time parameter `channels` is ignored and
+/// there's also a local temporary storage for a tap sum for each channel. This is faster,
+/// but requires extra registers for the intermediate sums.
+template <int static_channels = -1, typename Dst, typename Src>
+__device__ void ResampleHorz_Channels(
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_impl.cuh]: How is it different from what's done in dali/kernels/imgproc/resample/bilinear_impl.cuh ?

File: dali/kernels/imgproc/resample/resampling_internal_test.cu [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+  int W = img.shape[1];
+  int outH = H / 2;
+  auto gpu_mem_in = memory::alloc_unique<uint8_t>(AllocType::GPU, img.num_elements());
+  auto gpu_mem_out = memory::alloc_unique<uint8_t>(AllocType::GPU, outH * W * channels);
+
+  TensorView<StorageGPU, uint8_t, 3> img_in, img_out;
+  img_in = { gpu_mem_in.get(), img.shape };
+  img_out = { gpu_mem_out.get(), { outH, W, channels } };
+
+  copy(img_in, img);  // NOLINT
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_internal_test.cu]: why the NOLINT?

File: dali/kernels/imgproc/resample/resampling_internal_test.cu [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+    scale_x_ = static_cast<double>(out_w_) / InputWidth();
+    scale_y_ = static_cast<double>(out_h_) / InputWidth();
+    fixed_size_ = true;
+  }
+  void SetScale(double sx, double sy) {
+    scale_x_ = sx;
+    scale_y_ = sy;
+    out_w_ = sx * InputWidth();
+    out_h_ = sy * InputHeight();
+    fixed_size_ = false;  }
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/resampling_internal_test.cu]: break line here

File: dali/kernels/imgproc/resample/separable_impl_test.cc [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+  resampling.Run(ctx, out_tv, in_tv, params);
+  for (int i = 0; i < N; i++) {
+    auto ref_tensor = view_as_tensor<uint8_t, 3>(cv_ref[i]);
+    auto out_tensor = output.cpu()[i];
+    ASSERT_NO_FATAL_FAILURE(Check(out_tensor, ref_tensor, EqualEps(batch[i].epsilon)))
+    << [&]() {
+      cv::Mat tmp(out_tensor.shape[0], out_tensor.shape[1], CV_8UC3, out_tensor.data);
+      std::string inp_name = batch[i].input;
+      int ext = inp_name.rfind('.');
+#ifdef WINVER
```
Comment (jantonguirao) [inline dali/kernels/imgproc/resample/separable_impl_test.cc]: do we have any windows builds?

File: dali/pipeline/operators/resize/resize.cu [hunk 13]
```diff
@@ -115,16 +117,48 @@ void Resize<GPUBackend>::SetupSharedSampleParams(DeviceWorkspace* ws) {
...
+  auto interp_min = spec_.GetArgument<DALIInterpType>("min_filter");
+  auto interp_mag = spec_.GetArgument<DALIInterpType>("mag_filter");
+  auto interp_default = spec_.GetArgument<DALIInterpType>("interp_type");
+  if (interp_default != DALI_INTERP_LINEAR && interp_mag == DALI_INTERP_LINEAR)
+    interp_mag = interp_default;  // old syntax used
+
+  kernels::ResamplingFilterType interp2resample[] = {
+    kernels::ResamplingFilterType::Nearest,
+    kernels::ResamplingFilterType::Linear,
+    kernels::ResamplingFilterType::Linear,  // cubic - not supported
```
Comment (jantonguirao) [inline dali/pipeline/operators/resize/resize.cu]: should we warn the user at least?

File: dali/kernels/common/copy.h (hunk 1 above)
Comment (mzient) [reply dali/kernels/common/copy.h]: The view itself is const, the memory pointed to is not.

File: dali/kernels/imgproc/resample/resampling_internal_test.cu (hunk 10 above)
Comment (mzient) [reply dali/kernels/imgproc/resample/resampling_internal_test.cu]: cpplint had a problem with `copy` function supposedly requiring `#include <algorithm>`. cpplint has been modified.

File: dali/kernels/imgproc/resample/resampling_internal_test.cu (hunk 11 above)
Comment (mzient) [reply dali/kernels/imgproc/resample/resampling_internal_test.cu]: I wonder how it passed the linter O_o.

Comment (jantonguirao) [reply dali/kernels/imgproc/resample/resampling_internal_test.cu]: the linter doesn't seem very smart when it comes to indentations and stuff like that

File: dali/pipeline/operators/resize/resize.cu [hunk 14]
```diff
@@ -12,155 +12,105 @@
...
+  using Kernel = kernels::ResampleGPU<uint8_t, uint8_t>;
+  const auto &input = ws->Input<GPUBackend>(idx);
+  auto &kdata = kernel_data_.front();
+
+  kdata.context.gpu.stream = ws->stream();
+  kdata.requirements = Kernel::GetRequirements(
+      kdata.context,
+      view<const uint8_t, 3>(input),
+      resample_params_);
+  kdata.scratch_alloc.Reserve(kdata.requirements.scratch_sizes);
```
Comment (JanuszL) [inline dali/pipeline/operators/resize/resize.cu]: Can we avoid reallocating it for every run?
How expensive if reallocating it for every run?

Comment (mzient) [reply dali/pipeline/operators/resize/resize.cu]: `scratch_alloc` is persistent. It never shrinks.

Comment (JanuszL) [reply dali/pipeline/operators/resize/resize.cu]: It could. We should provide option to presize it as well.

**Comment by mzient** [comment]
CI passes: #654730

**Pull request merged**